---
title: "Differential Expression Stem Cells FMR1 WT/KO"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    se_file: "../data/se-2019-11-20.rds"
    design: !r formula(~ group)
    condition: "group"
    alpha: 0.1
    sig_genes_alpha: 0.05
    lfc: 1
    outputDir: "."
---

```{r, message = FALSE, warning = FALSE}
shhh <- suppressPackageStartupMessages
if(file.exists("_setup.R")){
    shhh(source("_setup.R"))
   
}

if(file.exists("_helperFunctions.R")){
    shhh(source("_helperFunctions.R"))
   
}
```

```{r header, child="_header.Rmd", eval=file.exists("_header.Rmd")}
```

```{r}
shhh(library(tidyverse))
shhh(library(DESeq2))
shhh(library(CHBUtils))
shhh(library(scales))
shhh(library(rio))
shhh(library(AnnotationHub))
shhh(library(DT))
shhh(library(DEGreport))
```

```{r helper_functions}
pca_loadings = function(object, ntop=nrow(object)) {
  rv <- matrixStats::rowVars(as.matrix(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop,
      length(rv)))]
  pca <- prcomp(t(as.matrix(object)[select, ]))
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  names(percentVar) = colnames(pca$x)
  pca$percentVar = percentVar
  return(pca)}

'%!in%' <- function(x,y)!('%in%'(x,y))
```

```{r}
set.seed(1234567890)
```

```{r}
# Directory paths
dataDir <- file.path(params$outputDir, "data",Sys.Date(),"stemcells_FMR1_WT_KO")
resultsDir <- file.path(params$outputDir, "results",Sys.Date(),"stemcells_FMR1_WT_KO")

deDir <- file.path(resultsDir, "differential_expression")
plotsDir <- file.path(resultsDir, "plots")
invisible(mapply(
    FUN = dir.create,
    list(dataDir, resultsDir, plotsDir, deDir),
    MoreArgs = list(recursive = TRUE, showWarnings = FALSE)
))
```

```{r}
se_dge <- readRDS(params$se_file)
```

```{r}
selected_samples <- colData(se_dge) %>% as.data.frame() %>% tibble::rownames_to_column(var = "sample_id") %>% dplyr::filter(class == "FMR1_WT_stemcell" | class == "FMR1_KO_stemcell") %>% dplyr::select(-c(batch,phenotype))

se_dge_selected <- se_dge[,selected_samples$sample_id]
```

```{r}
metrics_se_dge_selected <- se_dge_selected@metadata$metrics %>% tibble::remove_rownames()  %>% dplyr::filter(sample %in% selected_samples$sample_id) %>%  tibble::column_to_rownames(var = "sample")
metrics_se_dge_selected$sample_name <- rownames(metrics_se_dge_selected)
metadata_se_dge_selected <- colData(se_dge_selected) %>% as("data.frame") 
metadata_se_dge_selected$sample_name <- rownames(metadata_se_dge_selected)
counts_se_dge_selected <- assay(se_dge_selected)
summarydata_se_dge_selected <- metrics_se_dge_selected %>% dplyr::inner_join(metadata_se_dge_selected, by = "sample_name", keep = TRUE)
rownames(summarydata_se_dge_selected) <- summarydata_se_dge_selected$sample_name
counts_se_dge_selected <- counts_se_dge_selected[rowSums(counts_se_dge_selected)>0,]
counts_se_dge_selected <- counts_se_dge_selected[,rownames(summarydata_se_dge_selected)]

metadata_se_dge_selected <- droplevels(metadata_se_dge_selected)
summarydata_se_dge_selected <- droplevels(summarydata_se_dge_selected)

metadata_se_dge_selected$group <- relevel(metadata_se_dge_selected$group, ref = "FMR1_WT")
summarydata_se_dge_selected$group <- relevel(summarydata_se_dge_selected$group, ref = "FMR1_WT")
```

```{r}
ah <- AnnotationHub()
ahDb <- query(ah, pattern = c("Homo sapiens","EnsDb") )

id <- ahDb %>% 
     mcols() %>%  
        rownames() %>%
        tail(n = 1)

## Download the appropriate Ensembldb database
ahEnsdb <- ahDb[[id]]

rows <- genes(ahEnsdb)  %>% 
    as.data.frame() %>% 
    dplyr::mutate(gene_name = toupper(gene_name)) %>% 
    dplyr::select(c(gene_id,gene_name))
```

# Select data

We work with the dataset: stemcells-FMR1 WT/KO.

```{r}
if (file.exists(file.path(dataDir,paste0("dds-",Sys.Date(),".rds")))) {
    dds <- readRDS(file.path(dataDir,paste0("dds-stemcells_FMR1_WT_KO-",Sys.Date(),".rds")))
} else {
dds = DESeqDataSetFromMatrix(countData = counts_se_dge_selected, colData = summarydata_se_dge_selected, design = params$design)
dds = DESeq(dds)
saveRDS(dds,file = file.path(dataDir,paste0("dds-stemcells_FMR1_WT_KO-",Sys.Date(),".rds")))
}
```

```{r}
vst_dds = DESeq2::varianceStabilizingTransformation(dds)
vstMat <- assay(vst_dds)
```


# Results

We obtain the differential expression genes for the contrasts KO vs WT.

```{r}
all_results <- list()
contrast_string = paste("FMR1_KO_vs_FMR1_WT")
contrast = c("group", "FMR1_KO","FMR1_WT")
res_unshrunken <- DESeq2::results(
  dds,
  contrast = contrast,
  alpha = 0.1)
        
# DESeqResults with shrunken log2 fold changes (LFC).
res_shrunken <- lfcShrink(
dds = dds,
coef = 2,
type = 'apeglm',
res = res_unshrunken)
        
# Compare these two results and pick which one to report
res <- res_shrunken
        
res <- res[order(res$padj),]
res_unshrunken <- res_unshrunken[order(res_unshrunken$padj),]
res_shrunken <- res_shrunken[order(res_shrunken$padj),]
res_list = list("res_unshrunken" = res_unshrunken, "res_shrunken" = res_shrunken, "res" = res)
all_results[[length(all_results) + 1]] <- res_list
names(all_results) = contrast_string

```

## Plots

### Mean average (MA){.tabset}

MA-plots are a great way to visualize the comparative expression metrics for a two group comparison. The xâ€“axis is the average/mean expression over all the samples and the y-axis is the log2 fold change between the two conditions. The colored dots represent the genes that are differentially expressed (adjusted pvalue < `r params$sig_genes_alpha`). Upregulated genes are purple and downregulated, yellow.

We can observe a moderate number of differentially expressed genes.

```{r plot_ma, results = "asis", fig.height=10, fig.width=15}
for (i in seq(length(all_results))) {
  cat("\n###",names(all_results[i]),"\n")
  res = all_results[[i]]$res
  p <- plot_niceMA(res, alpha = params$sig_genes_alpha)
  print(p)
  ggsave(p, filename = file.path(plotsDir,paste0("maPlot-stemcells_",
                                                 names(all_results[i]),".png")))
  ggsave(p, filename = file.path(plotsDir,paste0("maPlot-stemcells_",
                                                 names(all_results[i]),".pdf")))
  ggsave(p, filename = file.path(plotsDir,paste0("maPlot-stemcells_",
                                                 names(all_results[i]),".tiff")))
  ggsave(p, filename = file.path(plotsDir,paste0("maPlot-stemcells_",
                                                 names(all_results[i]),".svg")))
  cat("\n")
}
```

## Volcano Plots {.tabset}

Volcano plots allow to visualize differential expression along to statistical significance. In the x-axis the log2 fold change between the two groups is represented. In the y-axis, the adjusted (for multiple testing) p-value for the test of Differential Expression. The green shade area surrounds genes with adj pval <`r params$sig_genes_alpha` and |logFC|> `r params$lfc`. 

As with the MA, we can see a moderate number of DE genes, with a higher number of them upregulated in the KO samples (positive log2FC).

```{r plot_volcano, results="asis"}
for (i in seq(length(all_results))) {
  cat("\n###",names(all_results[i]),"\n")
  res = all_results[[i]]$res
  stats = res %>% as.data.frame() %>% 
      dplyr::select(c(log2FoldChange,padj))
  p = volcano_density_plot(stats,  lfc.cutoff = params$lfc,pval.cutoff = params$sig_genes_alpha, title = names(all_results[i]))
  print(p)
  ggsave(p, filename = file.path(plotsDir,paste0("volcanoPlot-stemcells_",
                                                 names(all_results[i]),".png")))
  ggsave(p, filename = file.path(plotsDir,paste0("volcanoPlot-stemcells_",
                                                 names(all_results[i]),".pdf")))
  ggsave(p, filename = file.path(plotsDir,paste0("volcanoPlot-stemcells_",
                                                 names(all_results[i]),".tiff")))
  ggsave(p, filename = file.path(plotsDir,paste0("volcanoPlot-stemcells_",
                                                 names(all_results[i]),".svg")))
  cat("\n")
}
```


## Top significant genes expression.{.tabset}

In order to check the differential expression results, we plot the counts for the top significant genes. 

We can observe differences in the expression of genes from the contrasts.

```{r differentialPlot-topGenes, results="asis", message=FALSE}
rowData(dds)$ensgene <- rownames(dds)


rowData(dds) <- rowData(dds) %>% as.data.frame() %>% dplyr::left_join(rows,by = c("ensgene" = "gene_id")) %>% as("DataFrame")

for (i in seq(length(all_results))) {
  cat("\n####",names(all_results[i]),"\n")
  p = degPlot(dds,all_results[[i]]$res,n = min(sum(all_results[[i]]$res$padj <= params$sig_genes_alpha,na.rm = T),12),xs = "group",group = "group", ann  = c("ensgene","gene_name"))
  p <- p + ggtitle(names(all_results[i])) + ylab("counts (log2)")
  print(p)
  ggsave(p, file = file.path(plotsDir, paste0("degPlot-stemcells_",
                                              names(all_results[i]),".png")))
  ggsave(p, file = file.path(plotsDir, paste0("degPlot-stemcells_",
                                              names(all_results[i]),".pdf")))
  ggsave(p, file = file.path(plotsDir, paste0("degPlot-stemcells_",
                                              names(all_results[i]),".tiff")))
  ggsave(p, file = file.path(plotsDir, paste0("degPlot-stemcells_",
                                              names(all_results[i]),".svg")))
  cat("\n")
}
```

## Results table

```{r DESeq-tables, results='asis',echo=FALSE}
sig_genes_list = vector(mode = "list", length =
                            length(all_results))
sig_genes_list_names = vector(mode = "list", length =
                                  length(all_results))
for (i in seq(length(all_results))) {
  out_df = as.data.frame(all_results[[i]]$res)
  out_df$id = rownames(out_df)
  out_df = out_df[, c("id", colnames(out_df)[colnames(out_df) != "id"])]
  
  out_df <- out_df %>% dplyr::select(-c(baseMean,lfcSE)) %>% dplyr::left_join(rows, by = c("id" = "gene_id"))
  
  rio::export(out_df, file = file.path(deDir,paste0(names(all_results)[i], ".tsv")))
  sig_genes = subset(out_df, padj <= params$sig_genes_alpha)
  sig_genes_list[[i]] = sig_genes
  sig_genes_list_names[[i]] = sig_genes$id
  names(sig_genes_list)[i] = paste(names(all_results)[i])
  names(sig_genes_list_names)[i] = paste(names(all_results)[i])
  rio::export(sig_genes[order(sig_genes$padj),], file = file.path(deDir,paste0(names(all_results)[i],"_",params$sig_genes_alpha,"_sigGenes.tsv")))
}
```

For the contrast, two files are generated. The first one contains the results for all the genes and the second one, only for the significant ones (adjusted pval <`r params$sig_genes_alpha`).

Each file has 5 columns separated by tabs.

* id: Ensembl ID .
* log2FoldChange: Effect size estimate in the form of log2 fold change. For the form A vs B (in the file name), it represents the log2 of the ratio [expression in condition A]/[expression in condition B]. A positive value indicates that the gene is upregulated in condition A compared to B. A negative one, downregulated.
* pvalue: Wald test p-value.
* padj: BH adjusted p-values.
* Gene name/Symbol

### Example

```{r}
DT::datatable(head(out_df), style = "bootstrap")
```


# Significant genes

## Heatmap

### Heatmap with significant genes{.tabset}

This plot shows only differentially expressed genes on a per-sample basis. We have scaled the data by row.

We generate heatmaps for the different comparisons using the expression values of significant genes. Each row represents the normalized expression of a significant gene and each column, a sample. This plot helps to identify patterns of expression of genes between samples in the form of clusters. Expression of genes is centered and scaled by row to highlight differences in each gene sample. 


```{r DE-Heatmaps, results = "asis"}
tmp_metadata <- summarydata_se_dge_selected %>% as.data.frame() %>%  dplyr::select(c("group"))
vstMat <- assay(vst_dds)
for (i in seq(length(all_results))) {
  cat("\n####",names(all_results[i]),"\n")
  counts_SigGenes = vstMat[sig_genes_list[[i]]$id,]
  annNames <- rows %>% dplyr::filter(gene_id %in% sig_genes_list[[i]]$id) 
  rownames(annNames) <- annNames$gene_id
  if (length(counts_SigGenes) < 1){
    next
  }
  p <- pheatmap::pheatmap(counts_SigGenes, scale = "row", show_rownames = F, show_colnames = T, annotation_col = tmp_metadata, main = names(all_results[i]),
                     fontsize_row = 6, labels_row = annNames[sig_genes_list[[i]]$id,]$gene_name)
  print(p)
  ggsave(p, file = file.path(plotsDir, paste0("heatmap-stemcells_",
                                              names(all_results[i]),".png")))
  ggsave(p, file = file.path(plotsDir, paste0("heatmap-stemcells_",
                                              names(all_results[i]),".pdf")))
  ggsave(p, file = file.path(plotsDir, paste0("heatmap-stemcells_",
                                              names(all_results[i]),".tiff")))
  ggsave(p, file = file.path(plotsDir, paste0("heatmap-stemcells_",
                                              names(all_results[i]),".svg")))
  cat("\n")
}
```

## PCA 

### PCA with significant genes{.tabset}

We performed a PCA using the information from only the significant genes. We can observe that the significant genes clearly separate the compared samples in all the contrasts based on group(genotype).

Color is used to identify the group (genotype).

```{r }
colorby = "group"
shapeby = NULL
```

```{r pca , results="asis"}
for (i in seq(length(all_results))) {
  cat("\n####",names(all_results[i]),"\n")
  counts_SigGenes = vstMat[sig_genes_list[[i]]$id,]
  pc = pca_loadings(counts_SigGenes)
  comps = data.frame(pc$x)
  comps$fname = rownames(comps)
 comps <- left_join(comps,summarydata_se_dge_selected %>% as.data.frame() %>% tibble::rownames_to_column(var = "fname"),by = "fname")
  p <- customPlots::pca_plot_custom(pc,comps,1,2,colorby,shapeby)
  p <- p + ggtitle(names(all_results[i]))
  print(p)
  ggsave(p, file = file.path(plotsDir, paste0("pca-stemcells_",
                                              names(all_results[i]),".png")))
  ggsave(p, file = file.path(plotsDir, paste0("pca-stemcells_",
                                              names(all_results[i]),".pdf")))
  ggsave(p, file = file.path(plotsDir, paste0("pca-stemcells_",
                                              names(all_results[i]),".tiff")))
  ggsave(p, file = file.path(plotsDir, paste0("pca-stemcells_",
                                              names(all_results[i]),".svg")))
  cat("\n")
}
```

# Conclusions

We have performed a differential expression analysis and obtain a total of `r dim(sig_genes)[1]` genes with an adjusted pvalue less than `r params$sig_genes_alpha`. 

# Session 

```{r header, child="_footer.Rmd", eval=file.exists("_footer.Rmd")}
```